core/error.go

```diff
@@ -22,17 +22,40 @@ var (
 	// ErrKnownBlock is returned when a block to import is already known locally.
 	ErrKnownBlock = errors.New("block already known")
 
-	// ErrGasLimitReached is returned by the gas pool if the amount of gas required
-	// by a transaction is higher than what's left in the block.
-	ErrGasLimitReached = errors.New("gas limit reached")
-
 	// ErrBlacklistedHash is returned if a block to import is on the blacklist.
 	ErrBlacklistedHash = errors.New("blacklisted hash")
 
+	// ErrNoGenesis is returned when there is no Genesis Block.
+	ErrNoGenesis = errors.New("genesis not found in chain")
+)
+
+// State transition consensus errors, any of them encountered during
+// the block processing can lead to consensus issue.
+var (
+	// ErrNonceTooLow is returned if the nonce of a transaction is lower than the
+	// one present in the local chain.
+	ErrNonceTooLow = errors.New("nonce too low")
+
 	// ErrNonceTooHigh is returned if the nonce of a transaction is higher than the
 	// next one expected based on the local chain.
 	ErrNonceTooHigh = errors.New("nonce too high")
 
-	// ErrNoGenesis is returned when there is no Genesis Block.
-	ErrNoGenesis = errors.New("genesis not found in chain")
+	// ErrGasLimitReached is returned by the gas pool if the amount of gas required
+	// by a transaction is higher than what's left in the block.
+	ErrGasLimitReached = errors.New("gas limit reached")
+
+	// ErrInsufficientBalanceForTransfer is returned if the transaction sender doesn't
+	// have enough balance for transfer(topmost call only).
+	ErrInsufficientBalanceForTransfer = errors.New("insufficient balance for transfer")
+
+	// ErrInsufficientBalanceForFee is returned if transaction sender doesn't have
+	// enough balance to cover transaction fee.
+	ErrInsufficientBalanceForFee = errors.New("insufficient balance to pay fee")
+
+	// ErrGasOverflow is returned when calculating gas usage.
+	ErrGasOverflow = errors.New("gas overflow")
```

Comment:
I get a bit confused, because at another place (core/vm/errors.go) you have also defined `ErrGasUintOverflow`. So I guess two questions: 
- Why are some vm errors defined here, and some in core/vm/errors? What is the difference between the two categories? 
- Why do we have both `ErrGasUintOverflow` and `ErrGasOverflow`? Seems like it would be easy to mix those two up, and I'm not sure what (if any) the consequence would be?

Reply:
The errors defined here is for transaction pre-checking. If any error returned, it can lead to "BAD BLOCK".

The errors defined in `core/vm/errors` is for evm internal error. It's allowed behavior and will eventually be packed into `ExecutionResult`.

The former refers to some critical issues, the latter refers to evm execution status.
So it would be better to define them separately for better distinguish.

Reply:
I think the separation is a good idea, but maybe it could be made more obvious, for example if the errors in `core/vm` were called `EvmErrStackOverflow` instead of `ErrStackOverflow` ? 
I'm not sure..